WordPressAuthenticatorTests/Mocks/MockOnePasswordFacade.swift

```diff
@@ -0,0 +1,45 @@
+import Foundation
+@testable import WordPressAuthenticator
+
+class MockOnePasswordFacade: OnePasswordResultsFetcher {
+    var error: OnePasswordError?
+    var username: String?
+    var password: String?
+    var otp: String?
+    
+    init(username: String, password: String, otp: String?) {
+        self.error = nil
+        self.username = username
+        self.password = password
+        if let otp = otp {
+            self.otp = otp
+        } else {
+            self.otp = nil
+        }
+    }
+    
+    init(error: OnePasswordError) {
+        self.error = error
+        self.username = nil
+        self.password = nil
+        self.otp = nil
+    }
+    
+    
+    func findLogin(for url: String, viewController: UIViewController, sender: Any, success: @escaping (String, String, String?) -> Void, failure: @escaping (OnePasswordError) -> Void) {
+        if let error = error as NSError? {
+            failure(OnePasswordError(error: error))
+            return
+        }
+        
+        guard let username = username, let password = password else {
+            failure(.unknown)
+            return
+        }
+        
+        let oneTimePassword = otp
+        success(username, password, oneTimePassword)
```

Comment:
Because all the properties are optional, this function needs to do some checks to makes sure the mock hasn't been configured in an inconsistent way.

Swift's types system offers a few ways to _make inconsistent state unrepresentable_.

My favorite approach for a case like this would be to use `Result`.

```swift
// You can defined nested type in Swift. And if you make them private, they won't be visible 
// outside the parent type, which is handy to create little DTO objects like this one.
private struct Parameters {
  let username: String
  let password: String
  let otp: String?
}

// Using private, the users of this test double don't need to know how it works internally
private let result = Result<Parameters, OnePasswordError>

init(username: String, password: String, otp: String?) {
  result = .success(Parameters(username: username, password: password, otp: otp)
}

init(error: OnePasswordError) {
  result = .failure(error)
}

func findLogin(for url: String, viewController: UIViewController, sender: Any, success: @escaping (String, String, String?) -> Void, failure: @escaping (OnePasswordError) -> Void) {
  switch result {
    case .failure(let error): failure(error)
    case .success(let parameters): success(parameters.username, parameters.password, parameters.otp)
  }
}
```

_I haven't tried this code, it might need a few tweaks to compile._

Notice how with `Result` we remove the change in which some of the parameters are set but other s aren't and have to call `failure(.unknown)`, which would result in a confusing test failure.

Reply:
Thanks for this rework.  That's a much clearer way to approach it!